pb_decode.c

```diff
@@ -164,25 +155,72 @@ pb_istream_t pb_istream_from_buffer(const pb_byte_t *buf, size_t msglen)
     return stream;
 }
 
+/************************
+* Fast decode interface *
+*************************/
+
+#ifdef PB_ENABLE_MALLOC
+
+static void pb_decode_free(void *ptr)
+{
+    pb_free(ptr);
+}
+
+static void *pb_decode_realloc(void *ptr, size_t size)
+{
+    return pb_realloc(ptr, size);
+}
+
+#endif
+
+static void pb_decode_get_interface(pb_decode_if_t *pb) {
```

Comment:
I find this interface thing intriguing. On one hand it feels like it could have significant overhead, especially building it into RAM every time. On the other hand it could have some interesting optimization possibilities.

Could you elaborate why the functions are called through the interface pointers instead of directly?

Reply:
It was really just a workaround for how the tests are compiled right now.

The tests recompile pb_encode.c / pb_decode.c with different options enabled / disabled, and often then only link a binary with one or the other + the *.pb.o files. This makes it hard not only to just get the executable to link (e.g. when only pb_encode.o is linked), but means the *.pb.c files would also need to be recompiled when toggling malloc, double to float conversion, etc.

Exporting the interface means the *.pb.o code just ends up consuming whatever is in pb_decode.o - and it doesn't matter if it doesn't exist. I'd like to had a DYNAMIC / STATIC switch for the fast decode, where the STATIC which directly calls pb_*().

Reply:
It seems like this is just an issue of the test framework currently. I don't like such detail being frozen into the API design. I'd rather just have more limited test coverage for this new feature for the time being.